### DISCUSSION TITLE
String structure

### DISCUSSION THREAD

**Discussion by marartner** [vimeo/psalm#5278] [Ideas]
Psalm has certain pseudo-types for strings like `numeric-string`, that give a certain meaning to the string beyond being a mere string, or `lowercase-string` that enforces a certain format/structure. 

In my applications, I often pass around Uuids - generally as an object, which is no issue, but in some cases it has to be passed around as string, and I was imagining having a `uuid-string` psalm pseudo class would be cool for that - but that's probably too specific, and there's other things one might look for (`email-string`, `phone-string`, etc..) so I was wondering if it might be a worthwhile addition to generally be able to specify a string's structure, somewhat similar to how we already can specify an array structure; for example by adding a regex - something like `string</^[0-9a-f]{8}$/>`? - which we then could alias via `@psalm-type`:

`@psalm-type UuidString = string</^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i>`

Alternatively: is there a way to define our own pseudo-types via a plugin?

**Comment by weirdan** [comment]
For regexps, I don't think those would be useful to Psalm. And there's currently no way to define new scalar subtypes.

**Reply by orklah** [reply]
Oh? I wouldn't expect that to be an issue with a plugin? I suppose you should be able to create a class that extends TString and then override the of a method with ReturnTypeProviders for example, no?

EDIT: not the regex part, but the scalar subtype

**Reply by weirdan** [reply]
Yes and no. The type in Psalm is not self-contained, its interpretation is all over the place in Psalm internals. Now that you mentioned it, I think we discussed something similar here: https://github.com/vimeo/psalm/discussions/4826

**Reply by orklah** [reply]
Didn't he somewhat succeeded though? i'm not talking about creating a type out of thin air, but just to extends an existing type. This means it will behave exactly as it's super-type except when you want to flag it.

I see two potential issues:
- You can't pass this type to another scope (because it would need a change in parser to match the new type?)
- Transformations by Psalm may revert your type into a standard type at any time (this could be fixed in Psalm if needed I guess)

I will need to do something like that in the future, I'll be able to report the status on that then :)

**Comment by orklah** [comment]
The plugin route seems more natural for me.

One thing to keep in mind is that every new type in Psalm's core must be able to be compared to all others and to have rule for transformations

for example: 

```php
/** @var email-string $a */
$a = 'a@a.com';

$b = strtolower($a);
//what should be the type of $b here? email-string? lowercase-string?

$c = strreplace('.com', '.net', $a);
//and now? Is it still an email-string?

if($a === ''){
    //maybe we should define somewhere that an email-string can't ever be empty?
}

if(is_numeric($a)){
    //this shouldn't be possible either, but you have to say so somewhere
}
```

**Reply by marartner** [reply]
tbh for strtolower/strreplace I would have simply expected it to just return regular string and me having to tell it again that it's still an email-string via `@var` or so. no need to overcomplicate things. the other two cases are a good point though.
